File: benchmarks/live_suite/qsort.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

void qsort_custom(int64_t *arr, int low, int high) {
    if (low < high) {
        int64_t pivot = arr[high];
        int i = low;
        for (int j = low; j < high; j++) {
            if (arr[j] <= pivot) {
                int64_t temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
                i++;
            }
        }
        int64_t temp2 = arr[i];
        arr[i] = arr[high];
        arr[high] = temp2;

        int pi = i;
        if (pi > 0) qsort_custom(arr, low, pi - 1);
        qsort_custom(arr, pi + 1, high);
    }
}

int main(void) {
    int n = 100000;
    int64_t *arr = (int64_t *)malloc(n * sizeof(int64_t));
    for (int i = 0; i < n; i++) {
        arr[i] = ((int64_t)i * 1664525 + 1013904223) % 1000000;
    }
    qsort_custom(arr, 0, n - 1);
    printf("Qsort Checksum: %lld\n", (long long)arr[n / 2]);
    free(arr);
    return 0;
}

File: bench/live_suite/qsort.go
package main

import "fmt"

func qsortCustom(arr []int64, low, high int) {
	if low < high {
		pivot := arr[high]
		i := low
		for j := low; j < high; j++ {
			if arr[j] <= pivot {
				arr[i], arr[j] = arr[j], arr[i]
				i++
			}
		}
		arr[i], arr[high] = arr[high], arr[i]
		pi := i
		if pi > 0 {
			qsortCustom(arr, low, pi-1)
		}
		qsortCustom(arr, pi+1, high)
	}
}

func main() {
	n := 100000
	arr := make([]int64, n)
	for i := 0; i < n; i++ {
		arr[i] = (int64(i)*1664525 + 1013904223) % 1000000
	}
	qsortCustom(arr, 0, n-1)
	fmt.Printf("Qsort Checksum: %d\n", arr[n/2])
}
